### PR TITLE
fix(bootstrap): OOT credential provider support in 1.30+

### PR DESCRIPTION
**Pull request by Bryce-Soghigian**
**Description**

This fixes OOT credential provider support in AKS 1.30+:
* Uses correct download URL
* Pins down provider patch version
* Respects machine architecture

OOT credential provider should be supported in 1.29 as well, but this PR opts to not enable it just yet. We are likely to enable it some time later, after some more testing.

**How was this change tested?**

* `make presumbit`
* Local manual testing of bootstrap and image pull on 1.30.0; another test with faked detection of 1.30.2
* Retested on 1.28.9
* Local E2E Utilization tests (passed)
* CI E2E tests (most passed, Integration passed after retry, need to review for flakiness)

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note

```

### CODE REVIEW

**Review by Bryce-Soghigian — COMMENTED** [review]
/test

**Review by tallaxes — COMMENTED** [review]
/test

File: pkg/providers/imagefamily/bootstrap/aksbootstrap.go [hunk 1]
```diff
@@ -420,6 +420,28 @@ func kubeBinaryURL(kubernetesVersion, cpuArch string) string {
...
+		return ""
+	}
+
+	// credential provider has its own release outside of k8s version, and there'll be one credential provider binary for each k8s release,
+	// as credential provider release goes with cloud-provider-azure, not every credential provider release will be picked up unless
+	// there are CVE or bug fixes.
+	credentialProviderVersion := "1.29.2"
+	switch minorVersion {
+	case 29:
+		credentialProviderVersion = "1.29.2"
```
Comment (tallaxes) [inline pkg/providers/imagefamily/bootstrap/aksbootstrap.go:437]: Leaving this in as we would likely want to enable this for 1.29 as well, some time soon, after some more testing

**Comment by tallaxes** [comment]
@Bryce-Soghigian, please take a look as well, can't add you as reviewer since you initiated the PR

**Review by Bryce-Soghigian — COMMENTED** [review]
LGTM, I don't love how manual adding new selection for the binary URL is, but I can't think of a better way thats simpler.

File: pkg/providers/imagefamily/bootstrap/aksbootstrap.go [hunk 2]
```diff
@@ -420,6 +420,28 @@ func kubeBinaryURL(kubernetesVersion, cpuArch string) string {
...
+// or an empty string if OOT provider is not to be used
+func CredentialProviderURL(kubernetesVersion, arch string) string {
+	minorVersion := semver.MustParse(kubernetesVersion).Minor
+	if minorVersion < 30 { // use from 1.30; 1.29 supports it too, but we have not fully tested it with Karpenter
+		return ""
+	}
+
+	// credential provider has its own release outside of k8s version, and there'll be one credential provider binary for each k8s release,
+	// as credential provider release goes with cloud-provider-azure, not every credential provider release will be picked up unless
+	// there are CVE or bug fixes.
```
Comment (charliedmcb) [inline pkg/providers/imagefamily/bootstrap/aksbootstrap.go:433]: To clarify this comment. Each k8s version will have a binary. However, we are just linking to the minor version's binary, unless there are CVE, or bug fixes? Or will the released binary get replaced still getting us the CVE/bug fixes?

**Review by charliedmcb — APPROVED** [review]
Approved. However, have a few questions/comments.
 
In the PR description I just see how tested with "make presubmit". Feels worth testing as self hosted manually, or making sure we get successful e2e tests?

Also, left one clarify question on how the versioning actually works, and picking up of bug fixes/CVEs.

File: pkg/providers/imagefamily/bootstrap/aksbootstrap.go (hunk 2 above)
Comment (tallaxes) [reply pkg/providers/imagefamily/bootstrap/aksbootstrap.go:433]: This comment is copied from elsewhere, but my expectation is that, on CVEs and bug fixes, the next patch version of binary (for each affected minor version) will be released, and we will need to bump the patch version(s) here to pick it up.

**Comment by tallaxes** [comment]
> In the PR description I just see how tested with "make presubmit". Feels worth testing as self hosted manually, or making sure we get successful e2e tests?

Yes, doing more testing; will update the list

**Review by tallaxes — COMMENTED** [review]
/test

**Review by tallaxes — COMMENTED** [review]
/test

**Comment by tallaxes** [comment]
Most E2E tests passing in CI as well (intergration passed after retry, need to review for flakiness): https://github.com/Azure/karpenter-provider-azure/actions/runs/9966510663

**Comment by Bryce-Soghigian** [comment]
Lets get this in then?

**Pull request merged**